Iterate over dense dim first in split reduction reindexing

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147229

Fix for https://github.com/pytorch/pytorch/issues/144431.

Improves perf from 0.29963893827160504 -> 0.0396331632970453.

In split reductions, we view an input tensor as a single dimension, then reduce over it. When we are reducing over a tensor which has a dimension other than the last dimension as the dense dimension, we should iterate over the dense dimension first in our re-indexing. 

This pr also gives evidence for general need of reduction tiling, e.g. for cooperative reduction handling of this..


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

jansel reviewed (changes requested):
test failures?

shunting314 reviewed (commented):
Can we check cases like `a_col_major_2d_tensor.sum(dim=-1)`? (or more generally a reduction over multiple but not all dimensions of a multi-dimensional tensor) I'm worried if we may reorder dimensions for those non-reduction dimensions (which would be wrong).

eellison commented:
@shunting314 yes when I looked into regression on dashboard that was the cause. will update.

eellison committed:
Update on "Iterate over dense dim first in split reduction reindexing"


Fix for https://github.com/pytorch/pytorch/issues/144431.

Improves perf from 0.29963893827160504 -> 0.0396331632970453.

In split reductions, we view an input tensor as a single dimension, then reduce over it. When we are reducing over a tensor which has a dimension other than the last dimension as the dense dimension, we should iterate over the dense dimension first in our re-indexing. 

This pr also gives evidence for general need of reduction tiling, e.g. for cooperative reduction handling of this..


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Iterate over dense dim first in split reduction reindexing"


Fix for https://github.com/pytorch/pytorch/issues/144431.

Improves perf from 0.29963893827160504 -> 0.0396331632970453.

In split reductions, we view an input tensor as a single dimension, then reduce over it. When we are reducing over a tensor which has a dimension other than the last dimension as the dense dimension, we should iterate over the dense dimension first in our re-indexing. 

This pr also gives evidence for general need of reduction tiling, e.g. for cooperative reduction handling of this..


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -i
